File: control-plane-api/src/schemas/onboarding.py
"""Onboarding schemas — Pydantic v2 request/response models (CAB-1325)."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class OnboardingStepUpdate(BaseModel):
    """No body required — step name is in the URL path."""


class OnboardingProgressResponse(BaseModel):
    """Full onboarding progress response."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    user_id: str
    steps_completed: dict[str, str | None]
    started_at: datetime
    completed_at: datetime | None = None
    ttftc_seconds: int | None = None
    is_complete: bool = False


class OnboardingCompleteResponse(BaseModel):
    """Response from completing onboarding."""

    completed_at: datetime
    ttftc_seconds: int | None = None


class TrialKeyResponse(BaseModel):
    """Trial API key metadata (Phase 2 — not plaintext)."""

    key_prefix: str
    name: str
    rate_limit_rpm: int | None = None
    expires_at: datetime | None = None
    created_at: datetime


# Phase 3 — Analytics schemas

class FunnelStage(BaseModel):
    """Single stage in the onboarding funnel."""

    stage: str
    count: int
    conversion_rate: float | None = None


class FunnelResponse(BaseModel):
    """Full funnel analytics response."""

    stages: list[FunnelStage]
    total_started: int
    total_completed: int
    avg_ttftc_seconds: float | None = None
    p50_ttftc_seconds: float | None = None
    p90_ttftc_seconds: float | None = None


class StalledUserResponse(BaseModel):
    """User who started but hasn't completed onboarding."""

    user_id: str
    tenant_id: str
    last_step: str | None = None
    started_at: datetime
    hours_stalled: float

File: portal/src/services/onboarding.ts
/**
 * STOA Developer Portal - Onboarding Service (CAB-1325)
 *
 * API client for onboarding analytics endpoints (admin-only).
 */

import { apiClient } from './api';

// ============ Types ============

export interface FunnelStage {
  stage: string;
  count: number;
  conversion_rate: number | null;
}

export interface FunnelResponse {
  stages: FunnelStage[];
  total_started: number;
  total_completed: number;
  avg_ttftc_seconds: number | null;
  p50_ttftc_seconds: number | null;
  p90_ttftc_seconds: number | null;
}

export interface StalledUser {
  user_id: string;
  tenant_id: string;
  last_step: string | null;
  started_at: string;
  hours_stalled: number;
}

// ============ Service ============

export const onboardingService = {
  /**
   * Get onboarding funnel analytics
   * GET /v1/admin/onboarding/funnel
   */
  getFunnel: async (): Promise<FunnelResponse> => {
    const response = await apiClient.get<FunnelResponse>('/v1/admin/onboarding/funnel');
    return response.data;
  },

  /**
   * Get stalled users
   * GET /v1/admin/onboarding/stalled
   */
  getStalled: async (hours: number = 24): Promise<StalledUser[]> => {
    const response = await apiClient.get<StalledUser[]>('/v1/admin/onboarding/stalled', {
      params: { hours },
    });
    return response.data;
  },
};
